aioxmpp/stream.py

```diff
@@ -927,31 +927,46 @@ def _destroy_stream_state(self, exc):
             self.on_stream_destroyed(exc)
             self._established = False
 
-    def _iq_request_coro_done(self, request, task):
+    def _send_iq_reply(self, request, result):
+        if isinstance(result, errors.XMPPError):
+            response = request.make_reply(type_=structs.IQType.ERROR)
+            response.error = stanza.Error.from_exception(result)
+        else:
+            response = request.make_reply(type_=structs.IQType.RESULT)
+            response.payload = result
+        self._enqueue(response)
+
+    def _iq_request_coro_done_remove_task(self, task):
+        self._iq_request_tasks.remove(task)
+
+    def _iq_request_coro_done_send_reply(self, request, task):
         """
         Called when an IQ request handler coroutine returns. `request` holds
         the IQ request which triggered the excecution of the coroutine and
         `task` is the :class:`asyncio.Task` which tracks the running coroutine.
 
         Compose a response and send that response.
         """
-        self._iq_request_tasks.remove(task)
         try:
             payload = task.result()
         except errors.XMPPError as err:
-            response = request.make_reply(type_=structs.IQType.ERROR)
-            response.error = stanza.Error.from_exception(err)
+            self._send_iq_reply(request, err)
         except Exception:
             response = request.make_reply(type_=structs.IQType.ERROR)
             response.error = stanza.Error(
                 condition=errors.ErrorCondition.UNDEFINED_CONDITION,
                 type_=structs.ErrorType.CANCEL,
             )
+            self._enqueue(response)
             self._logger.exception("IQ request coroutine failed")
         else:
-            response = request.make_reply(type_=structs.IQType.RESULT)
-            response.payload = payload
-        self._enqueue(response)
+            self._send_iq_reply(request, payload)
```

Comment:
I’m wondering, we might actually want to split the composition of the IQ reply off of ``_send_iq_reply`` (into ``_compose_iq_reply``). Because composition can fail, and we must still send an IQ error (with undefined-condition) back in that case.

To distinguish cases where composition fails (when we can send undefined-condition back) and where enqueueing fails (where it is not likely that we can do anything useful), we need to split the function.